### PR TITLE
refactor(document): type annotate and refactor document reducers

### PR DESCRIPTION
**Pull request by rgbkrk**
Following in the foot steps of #1336 

Not ready yet.

### CODE REVIEW

**Comment by rgbkrk** [comment]
I really need to write up a manifesto on our state, so I'll do a light treatment here:

## Notebook Document

* Everything that eventually ends up persisted to the notebook document

## Transient, Associated with Kernel

* Display ID
* Comms, Models (ipywidgets)

## Transient, Associated with Document

* Cell view output

## Transient app state (local to single user)

* Focused cell
* Copied cell (?)

**Comment by rgbkrk** [comment]
For those following along, I'm amending to the same commit every time it's in a nice stable state rather than creating lots of commits. This is quite a heavy refactor to undertake.

**Comment by rgbkrk** [comment]
Created new notebooks, loaded old notebooks, and ran the conformance tests -- looking good from a compliance standpoint!

File: src/notebook/reducers/document.js [hunk 1]
```diff
@@ -47,325 +125,528 @@ export function cleanCellTransient(state, id) {
...
+function moveCell(state: DocumentState, action: MoveCellAction) {
+  return state.updateIn(['notebook', 'cellOrder'],
+    (cellOrder: CellOrder) => {
+      const oldIndex = cellOrder.findIndex(id => id === action.id);
+      const newIndex = cellOrder.findIndex(id => id === action.destinationId)
+                        + (action.above ? 0 : 1);
+      if (oldIndex === newIndex) {
+        return cellOrder;
+      }
+      return cellOrder
```
Comment (rgbkrk) [inline src/notebook/reducers/document.js:332]: This line is not covered by our current test suite.

File: src/notebook/reducers/document.js [hunk 2]
```diff
@@ -15,30 +84,39 @@ import * as constants from '../constants';
...
         nextToLast.get('output_type') === 'stream' &&
-        nextToLast.get('name') === output.name) {
-      return outputs.updateIn([outputs.size - 2, 'text'], text => text + output.text);
+        nextToLast.get('name') === streamOutput.name) {
+      return outputs.updateIn([outputs.size - 2, 'text'], appendText);
     }
   }
 
-  return outputs.push(Immutable.fromJS(output));
+  return outputs.push(Immutable.fromJS(streamOutput));
```
Comment (rgbkrk) [inline src/notebook/reducers/document.js:124]: This line is not covered by our current test suite.

File: src/notebook/reducers/document.js (hunk 1 above)
Comment (captainsafia) [reply src/notebook/reducers/document.js:332]: Yeah, this is that tricky one where we couldn't figure out exactly how it was behaving because it seemed like that line would never be hit based on the conditional and the way that `newIndex` and `oldIndex` were set.

File: src/notebook/reducers/document.js [hunk 3]
```diff
@@ -47,325 +125,528 @@ export function cleanCellTransient(state, id) {
...
-    }
+function setNotebook(state: DocumentState, action: SetNotebookAction) {
+  const notebook = action.notebook
+    .update('cellMap', (cells: CellMap): CellMap =>
+      // TODO: Determine why setting the notebook is changing these values.
+      //       Should they be transient or part of an overall app state?
+      cells.map(value =>
+        value.setIn(['metadata', 'inputHidden'], false)
+              .setIn(['metadata', 'outputHidden'], false)
+              .setIn(['metadata', 'outputExpanded'], false)));
```
Comment (rgbkrk) [inline src/notebook/reducers/document.js:150]: @captainsafia do we want to be persisting `outputHidden` to the notebooks or is it only for at runtime? The fact that these get cleaned up on load seems like we're currently:


```
* Setting metadata on the notebooks
* Saving with the metadata (woo my hidden setting is acknowledged)
* Resetting them on load?
```

File: src/notebook/reducers/document.js (hunk 1 above)
Comment (rgbkrk) [reply src/notebook/reducers/document.js:332]: When a user is using this, it must be getting hit because otherwise cells would stay fixed in the prior order. 🤔

File: src/notebook/reducers/document.js (hunk 3 above)
Comment (captainsafia) [reply src/notebook/reducers/document.js:150]: These values are currently persisted to the notebook on save. The reason we didn't do this was because there was some discussion about persisting arbitrary (not defined in the notebook format) values. So as of now, this configuration is only used during runtime and forgotten once the user closes out of nteract.

Comment (rgbkrk) [reply src/notebook/reducers/document.js:150]: Ok, cool thanks for clarifying it! This helps me see why these are being set to false here. It isn't to wipe out what the user may have had set in the notebook, it's for the structure to be available within this cell when rendered. The colocation of the data here is for our internal use (it makes sense to have it as part of the cell), and not persisted to the document.

Next steps as we start refactoring state to scope is to figure out if we:

* Keep state like this colocated with the cell, initialize on load, delete/sanitize before save
* Create a separate state structure for transient-at-document-runtime (as opposed to the current transient which is transient-by-kernel-runtime)

Comment (rgbkrk) [reply src/notebook/reducers/document.js:150]: > there was some discussion about persisting arbitrary (not defined in the notebook format) values

Is this a feature where we should persist them and use our own prefix:

```
metadata: {
  "nteractOutputHidden": ...
}
```

We would definitely be on the hook for maintaining this as our own spec (including migrations in the future).

Comment (rgbkrk) [reply src/notebook/reducers/document.js:150]: Where we've run into issues before with not conforming to the spec was when keys were set above metadata, which [does allow for additional properties](https://github.com/jupyter/nbformat/blob/62d6eb8803616d198eaa2024604d1fe923f2a7b3/nbformat/v4/nbformat.v4.schema.json#L175).

File: src/notebook/reducers/document.js (hunk 1 above)
Comment (rgbkrk) [reply src/notebook/reducers/document.js:332]: I've submitted a test that covers this now, it seemed important enough to make sure we reach it with enough edge cases.

**Comment by rgbkrk** [comment]
This should now be fully covered after the next test + codecov run!

**Comment by peggyrayzis** [comment]
Feels awesome to merge this PR! Thanks for getting the ball rolling w/ typing the reducers 👍

**Comment by jdetle** [comment]
Very leet PR here.

**Comment by rgbkrk** [comment]
> Thanks for getting the ball rolling w/ typing the reducers

My pleasure, I'm glad we took the time to do this. Helped me see a lot of the state we need to clean up.
